examples/bri-3/src/bri/communication/agents/bpiMessages.agent.ts

```diff
@@ -1,31 +1,101 @@
-import { Injectable, NotFoundException } from '@nestjs/common';
+import {
+  BadRequestException,
+  Injectable,
+  NotFoundException,
+} from '@nestjs/common';
+import { LoggingService } from '../../../shared/logging/logging.service';
+import { BpiSubjectStorageAgent } from '../../identity/bpiSubjects/agents/bpiSubjectsStorage.agent';
+import { BpiSubject } from '../../identity/bpiSubjects/models/bpiSubject';
+import {
+  BPI_MESSAGE_ALREADY_EXISTS,
+  NOT_FOUND_ERR_MESSAGE,
+} from '../api/err.messages';
 import { BpiMessage } from '../models/bpiMessage';
 import { BpiMessageType } from '../models/bpiMessageType.enum';
-import { NOT_FOUND_ERR_MESSAGE } from '../api/err.messages';
 import { BpiMessageStorageAgent } from './bpiMessagesStorage.agent';
-import { BpiSubject } from '../../identity/bpiSubjects/models/bpiSubject';
-import { BpiSubjectStorageAgent } from '../../identity/bpiSubjects/agents/bpiSubjectsStorage.agent';
 
 @Injectable()
 export class BpiMessageAgent {
   constructor(
     private bpiMessageStorageAgent: BpiMessageStorageAgent,
     private bpiSubjectStorageAgent: BpiSubjectStorageAgent,
+    private readonly logger: LoggingService,
   ) {}
 
-  public async getFromAndToSubjectsAndThrowIfNotExist(
+  public async validateNewBpiMessageAgainstExistingBpiEntitiesWithThrow(
+    messageId: string,
     fromBpiSubjectId: string,
     toBpiSubjectId: string,
   ): Promise<[BpiSubject, BpiSubject]> {
+    const existingBpiMessage =
+      await this.bpiMessageStorageAgent.getBpiMessageById(messageId);
+
+    if (existingBpiMessage) {
+      throw new BadRequestException(BPI_MESSAGE_ALREADY_EXISTS);
+    }
+
     const fromBpiSubject = await this.bpiSubjectStorageAgent.getBpiSubjectById(
       fromBpiSubjectId,
     );
+
+    if (!fromBpiSubject) {
+      throw new NotFoundException(NOT_FOUND_ERR_MESSAGE);
+    }
+
     const toBpiSubject = await this.bpiSubjectStorageAgent.getBpiSubjectById(
       toBpiSubjectId,
     );
+
+    if (!toBpiSubject) {
+      throw new NotFoundException(NOT_FOUND_ERR_MESSAGE);
+    }
+
     return [fromBpiSubject, toBpiSubject];
   }
 
+  public async validateNewBpiMessageAgainstExistingBpiEntities(
```

Comment:
method name and purpose a bit unclear, it is doing same thing as the one above, but also returning bpi subjects?

Reply:
Yes, this one differs from the one above because it does not throw a HTTP exception in case something not found. I do not like it as well, it is doing to many things so cannot be named precisely. I ll try to split it and rename.